File: backend/app/decompiler.py
import os
import sys
import time
import shutil
import zipfile
import subprocess
from pathlib import Path
from typing import Optional, List

from backend.app.models import DecompileStats

TIMEOUT_SECONDS = 90

def find_jadx_binary() -> Optional[str]:
    """
    Finds jadx binary from bundled tools/jadx or system PATH.
    """
    # 1. Check bundled tools/jadx/bin relative to project root
    project_root = Path(__file__).resolve().parent.parent.parent
    bundled_dir = project_root / "tools" / "jadx" / "bin"
    
    if sys.platform.startswith("win"):
        bundled_exe = bundled_dir / "jadx.bat"
    else:
        bundled_exe = bundled_dir / "jadx"
        
    if bundled_exe.is_file():
        return str(bundled_exe)
        
    # 2. Check system PATH
    system_jadx = shutil.which("jadx")
    if system_jadx:
        return system_jadx
        
    return None

def count_java_files(directory: Path) -> int:
    """Counts total .java files in directory tree quickly using os.walk."""
    if not directory.exists():
        return 0
    count = 0
    for _, _, files in os.walk(str(directory)):
        for f in files:
            if f.endswith(".java") or f.endswith(".smali"):
                count += 1
    return count

def _extract_raw_java_source(apk_path: Path, output_dir: Path) -> int:
    """
    Fallback: extracts .java files if the APK archive is a source bundle.
    """
    count = 0
    sources_dir = output_dir / "sources"
    with zipfile.ZipFile(apk_path, "r") as z:
        for name in z.namelist():
            if name.endswith(".java"):
                # Avoid zip-slip
                p = Path(name)
                dest = sources_dir / p
                dest.parent.mkdir(parents=True, exist_ok=True)
                with z.open(name) as src, open(dest, "wb") as dst:
                    dst.write(src.read())
                count += 1
    return count

def decompile_apk(apk_path: Path, job_dir: Path, timeout: int = TIMEOUT_SECONDS) -> DecompileStats:
    """
    Decompiles an APK using jadx into job_dir/decompiled.
    Handles timeout, failures, partial decompilations, and fallback for source-bundle APKs.
    Surfaces decompilation_incomplete and decompilation_warnings in DecompileStats.
    """
    decompiled_dir = job_dir / "decompiled"
    decompiled_dir.mkdir(parents=True, exist_ok=True)
    
    start_time = time.time()
    jadx_bin = find_jadx_binary()
    
    if not jadx_bin:
        # If jadx is missing, try raw source fallback before failing
        extracted = _extract_raw_java_source(apk_path, decompiled_dir)
        elapsed = round(time.time() - start_time, 2)
        if extracted > 0:
            return DecompileStats(
                status="complete",
                method="raw_source",
                file_count=extracted,
                time_taken_seconds=elapsed,
                decompilation_incomplete=False,
                decompilation_warnings=[]
            )
        return DecompileStats(
            status="failed",
            method="none",
            file_count=0,
            time_taken_seconds=elapsed,
            error="jadx binary not found. Please run setup_jadx.ps1 or setup_jadx.sh.",
            decompilation_incomplete=True,
            decompilation_warnings=["jadx binary not found on system PATH or in tools/jadx"]
        )

    # Invoke jadx CLI with memory-capped and I/O-optimized flags
    cmd = [
        jadx_bin,
        "-j", "2",          # Limit to 2 threads to prevent NVMe/SSD saturation and 100% active disk queue
        "-d", str(decompiled_dir),
        "--no-res",         # Skip resources to speed up code extraction
        "--no-debug-info",  # Skip debug line/var tables, reducing memory and disk writes by ~40%
        str(apk_path)
    ]
    
    # Restrict JVM Heap to 768 MB so it never hogs host system memory
    sub_env = os.environ.copy()
    sub_env["JAVA_OPTS"] = "-Xmx768m -Xms128m -XX:+UseG1GC"
    sub_env["DEFAULT_JVM_OPTS"] = "-Xmx768m -Xms128m -XX:+UseG1GC"

    try:
        proc = subprocess.run(
            cmd,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
            timeout=timeout,
            env=sub_env
        )
        elapsed = round(time.time() - start_time, 2)
        file_count = count_java_files(decompiled_dir)
        
        warnings: List[str] = []
        stderr_text = proc.stderr or ""
        stdout_text = proc.stdout or ""
        combined_output = f"{stdout_text}\n{stderr_text}"
        
        # Check for errors/warnings in output
        for line in combined_output.splitlines():
            line_str = line.strip()
            if "ERROR" in line_str or "WARN" in line_str:
                if len(line_str) > 200:
                    line_str = line_str[:200] + "..."
                if line_str not in warnings:
                    warnings.append(line_str)
                    if len(warnings) >= 10:  # limit noise
                        break

        incomplete = False
        if proc.returncode != 0:
            incomplete = True
            warnings.append(f"jadx process exited with non-zero code {proc.returncode}")

        if file_count > 0:
            return DecompileStats(
                status="complete",
                method="jadx",
                file_count=file_count,
                time_taken_seconds=elapsed,
                decompilation_incomplete=incomplete or len(warnings) > 0,
                decompilation_warnings=warnings
            )
            
        # If jadx exited with 0 files, check if APK has raw .java files (e.g. test fixture archive)
        extracted = _extract_raw_java_source(apk_path, decompiled_dir)
        if extracted > 0:
            return DecompileStats(
                status="complete",
                method="raw_source",
                file_count=extracted,
                time_taken_seconds=elapsed,
                decompilation_incomplete=False,
                decompilation_warnings=[]
            )
            
        err_msg = proc.stderr.strip() or proc.stdout.strip() or "jadx produced no Java files."
        return DecompileStats(
            status="failed",
            method="jadx",
            file_count=0,
            time_taken_seconds=elapsed,
            error=err_msg[:500],
            decompilation_incomplete=True,
            decompilation_warnings=[err_msg[:300]]
        )
        
    except subprocess.TimeoutExpired:
        elapsed = round(time.time() - start_time, 2)
        file_count = count_java_files(decompiled_dir)
        return DecompileStats(
            status="failed" if file_count == 0 else "complete",
            method="jadx",
            file_count=file_count,
            time_taken_seconds=elapsed,
            error=f"Decompilation timed out after {timeout} seconds.",
            decompilation_incomplete=True,
            decompilation_warnings=[f"Decompilation timed out after {timeout} seconds; partial files may have been recovered."]
        )
    except Exception as e:
        elapsed = round(time.time() - start_time, 2)
        return DecompileStats(
            status="failed",
            method="jadx",
            file_count=0,
            time_taken_seconds=elapsed,
            error=str(e),
            decompilation_incomplete=True,
            decompilation_warnings=[str(e)]
        )

File: src/lib/analyzer/decompiler.ts
import fs from "fs";
import path from "path";
import { execFile } from "child_process";
import AdmZip from "adm-zip";
import { DecompileStats } from "./types";
import { CodeFile } from "./rules-engine";
import { extractDexStrings } from "./dex-extractor";

export function findJadxBinary(): string | null {
  const projectRoot = process.cwd();
  const isWin = process.platform === "win32";

  // 1. Check tools/jadx/bin
  const bundledBat = path.join(projectRoot, "tools", "jadx", "bin", "jadx.bat");
  const bundledSh = path.join(projectRoot, "tools", "jadx", "bin", "jadx");

  if (isWin && fs.existsSync(bundledBat)) {
    return bundledBat;
  }
  if (!isWin && fs.existsSync(bundledSh)) {
    return bundledSh;
  }

  // 2. Check system PATH
  const pathDirs = (process.env.PATH || "").split(path.delimiter);
  const exeName = isWin ? "jadx.bat" : "jadx";
  for (const dir of pathDirs) {
    const candidate = path.join(dir, exeName);
    if (fs.existsSync(candidate)) {
      return candidate;
    }
  }

  return null;
}

function collectFilesRecursively(dir: string, baseDir: string): CodeFile[] {
  const files: CodeFile[] = [];
  if (!fs.existsSync(dir)) return files;

  const entries = fs.readdirSync(dir, { withFileTypes: true });
  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      files.push(...collectFilesRecursively(fullPath, baseDir));
    } else if (entry.isFile() && (entry.name.endsWith(".java") || entry.name.endsWith(".kt") || entry.name.endsWith(".smali"))) {
      try {
        const content = fs.readFileSync(fullPath, "utf8");
        const relPath = path.relative(baseDir, fullPath).replace(/\\/g, "/");
        files.push({ relative_path: relPath, content });
      } catch {
        // ignore unreadable files
      }
    }
  }
  return files;
}

export async function decompileAndExtractCode(
  apkPath: string,
  jobDir: string,
  zip: AdmZip
): Promise<{ stats: DecompileStats; codeFiles: CodeFile[] }> {
  const startTime = Date.now();
  const decompiledDir = path.join(jobDir, "decompiled");
  if (!fs.existsSync(decompiledDir)) {
    fs.mkdirSync(decompiledDir, { recursive: true });
  }

  const jadxBin = findJadxBinary();

  // 1. Try JADX execution if binary is present
  if (jadxBin) {
    try {
      const stats = await new Promise<DecompileStats>((resolve) => {
        const args = ["-d", decompiledDir, "--no-res", apkPath];
        const child = execFile(jadxBin, args, { timeout: 90000, shell: true }, (error, stdout, stderr) => {
          const elapsed = Number(((Date.now() - startTime) / 1000).toFixed(2));
          const warnings: string[] = [];
          const combined = `${stdout || ""}\n${stderr || ""}`;

          for (const line of combined.split("\n")) {
            const trimmed = line.trim();
            if (trimmed.includes("WARN") || trimmed.includes("ERROR")) {
              if (warnings.length < 10) {
                warnings.push(trimmed.slice(0, 200));
              }
            }
          }

          if (error && !fs.existsSync(decompiledDir)) {
            resolve({
              status: "failed",
              method: "jadx",
              file_count: 0,
              time_taken_seconds: elapsed,
              error: error.message,
              decompilation_incomplete: true,
              decompilation_warnings: [error.message],
            });
            return;
          }

          const files = collectFilesRecursively(decompiledDir, decompiledDir);
          resolve({
            status: files.length > 0 ? "complete" : "partial",
            method: "jadx",
            file_count: files.length,
            time_taken_seconds: elapsed,
            decompilation_incomplete: Boolean(error) || warnings.length > 0,
            decompilation_warnings: warnings,
          });
        });
      });

      const files = collectFilesRecursively(decompiledDir, decompiledDir);
      if (files.length > 0) {
        return { stats, codeFiles: files };
      }
    } catch (e: any) {
      console.warn("Jadx execution failed, falling back to archive extraction", e);
    }
  }

  // 2. Fallback: Extract raw .java / .kt files from zip
  const rawCodeFiles: CodeFile[] = [];
  const zipEntries = zip.getEntries();
  for (const entry of zipEntries) {
    if (
      !entry.isDirectory &&
      (entry.entryName.endsWith(".java") || entry.entryName.endsWith(".kt") || entry.entryName.endsWith(".smali"))
    ) {
      try {
        const content = entry.getData().toString("utf8");
        rawCodeFiles.push({
          relative_path: entry.entryName.replace(/\\/g, "/"),
          content,
        });
      } catch {
        // ignore
      }
    }
  }

  if (rawCodeFiles.length > 0) {
    const elapsed = Number(((Date.now() - startTime) / 1000).toFixed(2));
    return {
      stats: {
        status: "complete",
        method: "raw_source",
        file_count: rawCodeFiles.length,
        time_taken_seconds: elapsed,
        decompilation_incomplete: false,
        decompilation_warnings: [],
      },
      codeFiles: rawCodeFiles,
    };
  }

  // 3. DEX Bytecode String Extraction Fallback
  const dexCodeFiles: CodeFile[] = [];
  let dexCount = 0;
  for (const entry of zipEntries) {
    if (!entry.isDirectory && entry.entryName.endsWith(".dex")) {
      dexCount++;
      const dexBuf = entry.getData();
      const extractedStrings = extractDexStrings(dexBuf);
      if (extractedStrings.length > 0) {
        dexCodeFiles.push({
          relative_path: entry.entryName,
          content: extractedStrings.join("\n"),
        });
      }
    }
  }

  const elapsed = Number(((Date.now() - startTime) / 1000).toFixed(2));

  if (dexCodeFiles.length > 0) {
    return {
      stats: {
        status: "complete",
        method: "dex_strings",
        file_count: dexCodeFiles.length,
        time_taken_seconds: elapsed,
        decompilation_incomplete: false,
        decompilation_warnings: jadxBin
          ? []
          : ["JADX not available; analyzed raw DEX bytecode string constants."],
      },
      codeFiles: dexCodeFiles,
    };
  }

  return {
    stats: {
      status: "failed",
      method: "none",
      file_count: 0,
      time_taken_seconds: elapsed,
      error: "No source files or DEX bytecode could be extracted.",
      decompilation_incomplete: true,
      decompilation_warnings: ["No extractable code files found in archive."],
    },
    codeFiles: [],
  };
}
